sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -180,6 +180,70 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
   return httpRequest;
 }
+
+static void CleanupThread()
+{
+  using namespace Azure::Core::Http::_detail;
+  for (;;)
+  {
+    {
+      // Won't continue until the ConnectionPoolMutex is released from MoveConnectionBackToPool
+      std::unique_lock<std::mutex> lockForPoolCleaning(
+          CurlConnectionPool::g_curlConnectionPool.ConnectionPoolMutex);
+      // Wait the defined default time OR to the signal from the conditional variable.
+      // wait_until releases the mutex lock until it wakes up again or it's cancelled.
+      if (CurlConnectionPool::g_curlConnectionPool.ConditionalVariableForCleanThread.wait_for(
+              lockForPoolCleaning,
+              std::chrono::milliseconds(DefaultCleanerIntervalMilliseconds),
+              []() { return CurlConnectionPool::g_curlConnectionPool.ConnectionCounter == 0; }))
+      {
+        // Cancelled by another thead or no connections on wakeup
+        return;
+      }
+
+      // loop the connection pool index - Note: lock is re-taken for the mutex
+      for (auto index = CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.begin();
+           index != CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.end();
+           index++)
+      {
+        if (index->second.size() == 0)
+        {
+          // Move the next pool index
+          continue;
+        }
+
+        // Pool index with waiting connections. Loop the connection pool backwards until
+        // a connection that is not expired is found or until all connections are removed.
```

Comment:
Why backwards?

Reply:
Improving comment to:

// Pool index with waiting connections. Each pool index behaves as a Last-in-First-out list,
      // the last connection moved to the pool will be the first to be re-used. Because of this, the
      // oldest connection in the pool can be found at the end of the list. Looping the connection
      // pool backwards until a connection that is not expired is found or until all connections are
      // removed.

Reply:
>  Because of this, the oldest connection in the pool can be found at the end of the list.

Is that true? In a LIFO, wouldn't the newest connection be at the end, while the oldest one would be at the beginning? What do you mean by oldest?

> Looping the connection pool backwards until a connection that is not expired is found or until all connections are
 removed.

It's still not clear to me **why** we loop backwards, why we want LIFO behavior. What's the benefit exactly? That's what I would love to see as a comment.

Reply:
When I created the design for the pool, I had to decide as a connection pool re-using a connection, which should I re-use first (if I have more than one). The newest (the last connection moved to the pool) or the oldest (the first connection that came to the pool).  
After taking to the team, we decide that there are more chances of success for the newest connection to work than for the oldest. So the pool is similar to a stack, we push and pop from the top and let the connections which came fist to get older and older on the bottom.  
That's why, by checking the connections from the bottom to the top, as soon as one connection is not expired, is a sentinel to know that all other connections (that are newer) are not expired as well.

Reply:
> we decide that there are more chances of success for the newest connection to work than for the oldest.

That's a pretty good rationale. Makes sense!

> That's why, by checking the connections from the bottom to the top, as soon as one connection is not expired, is a sentinel to know that all other connections (that are newer) are not expired as well.

I am now confused about our orientation and what you mean by bottom and top. In a LIFO stack, the last item is on the top, so I thought we'd check the top first (not bottom).

By bottom, I assume you mean end of the list, right? So we go from the end of the list to the beginning of the list. How is the first item in the list newer than the second or last? Also, what do you mean by "a connection expiring"?